### PR TITLE
feat: support create/update of embedded JSON objects

### PR DESCRIPTION
**Pull request by craicoverflow**
This adds support for create/update of embedded relationship documents in both PostgreSQL and MongoDB.

Before this change, it was not possible to CREATE/UPDATE/FIND fields which are GraphQL types which are not models because:

a) There were no input fields added for these types to the CRUD input types
b) `getSelectedFieldsFromResolverInfo` did not include GraphQL type fields in the selectable fields, so `.select(x)` would never return this data from the database.

TL;DR we added support for non-model type fields in out models in https://github.com/aerogear/graphback/issues/1432 but this data could not be inserted/updated/retrieved from the database.

```graphql
"""
@model
"""
type Note {
  id: ID!
  title: String
  """
  this annotation is only needed for Postgres
  @db(type: 'json') 
  """
  comments: [Comment]
}

type Comment {
  owner: String
  comment: String
}
```

You can perform a mutation like so:

```gql
mutation {
  createNote(input: {
    title:"my new note",
    comments: [
      {
        comment:"Awesome!",
        owner:"Enda"
      }
    ]
  }) {
    id
    title
    comments {
      owner
      comment
    }
  }
}
```

### Tasks

- [x] Create input types for non-model fields
- [x] Allow CREATE/UPDATE/FIND on non-model fields (Mongo/Postgres)
- [x] Integration tests

### CODE REVIEW

**Comment by machi1990** [comment]
You rock @craicoverflow, looking at this now!

**Comment by craicoverflow** [comment]
It is not possible to query the embedded relationship using Postgres with Knex (unless you do raw queries, which are not a quick fix)

```
findNotes(filter: {comment: {contains: "{hello: 1}"}}) {
    items {
      comment {
        description
        test {
          hello
        }
      }
    }
 }
```

This throws: `"select \"comment\" from \"note\" where \"comment\" like $1 - operator does not exist: json ~~ unknown""`

I am now going to do some testing with Mongo, and see if it works. We can decide whether to disable filtering on both for now until we have a solution for Postgres.

**Comment by craicoverflow** [comment]
Filtering on the field for Mongo does nothing either, going to disable that.

**Comment by wtrocki** [comment]
Did I missed some issue. Trying to recall when we talked about this being problematic etc?

**Comment by craicoverflow** [comment]
@wtrocki see https://github.com/aerogear/graphback/issues/1540

**Comment by wtrocki** [comment]
> It is not possible to query the embedded relationship using Postgres with Knex (unless you do raw queries, which are not a quick fix)


But aren't this coming from graphql resolvers ?

**Comment by wtrocki** [comment]
What happening here - are we downgrading graphqlcrud features? Tell me like I'm 5 y old :D

**Comment by craicoverflow** [comment]
The data exists in the database so it would be more performant to filter the JSON or embedded document in the database.

**Comment by craicoverflow** [comment]
The CRUD has been upgraded - no features have been removed, because now it supports strongly-typed fields which are non-model objects.

Before this change, it was not possible to CREATE/UPDATE/FIND fields which are GraphQL types which are not models because:

a) There were no input fields added for these types to the CRUD input types
b) `getSelectedFieldsFromResolverInfo` did not include GraphQL type fields in the selectable fields, so `.select(x)` would never return this data from the database.

TL;DR we added support for non-model type fields in out models in https://github.com/aerogear/graphback/issues/1432 but this data could not be inserted/updated/retrieved from the database.

**Comment by craicoverflow** [comment]
@wtrocki @machi1990 I need to add tests, but ready for review other than that.

**Comment by wtrocki** [comment]
Mind blown! Calling out MachiMan hero to help here(@machi1990) 
I will need to have more time to understand this better.

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts [hunk 1]
```diff
@@ -258,8 +281,64 @@ export const buildMutationInputType = (modelType: GraphQLObjectType) => {
...
+  return fields.map((field: GraphQLField<any, any>) => {
+    let namedType = getNamedType(field.type)
+    let typeName = namedType.name
+
+    let inputType
+    if (isObjectType(namedType)) {
+      typeName = getInputTypeName(typeName, GraphbackOperationType.CREATE)
+      namedType = schemaComposer.getOrCreateITC(typeName).getType()
+
+      inputType = copyWrappingType(field.type, namedType)
```
Comment (machi1990) [inline packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:298]: Why is copyingWrapping type needed?

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts [hunk 2]
```diff
@@ -258,8 +281,64 @@ export const buildMutationInputType = (modelType: GraphQLObjectType) => {
...
+  schemaComposer.add(mutationInputObject)
+}
+
+function mapObjectInputFields(schemaComposer: SchemaComposer<any>, fields: GraphQLField<any, any>[]): GraphQLInputField[] {
+  return fields.map((field: GraphQLField<any, any>) => {
+    let namedType = getNamedType(field.type)
+    let typeName = namedType.name
+
+    let inputType
+    if (isObjectType(namedType)) {
```
Comment (machi1990) [inline packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:294]: What happens if the type is a model?

File: packages/graphback-runtime-knex/src/knexQueryMapper.ts [hunk 3]
```diff
@@ -101,7 +101,7 @@ function where(builder: Knex.QueryBuilder, filter: any, or: boolean = false, not
 
     // eslint-disable-next-line no-null/no-null
     if (exprEntry[1] === null) {
-      builder = builder[builderMethod('whereNull', or, not)](col)
+      builder = builder[builderMethod('whereNull', or, exprEntry[0] === 'ne')](col)
```
Comment (machi1990) [inline packages/graphback-runtime-knex/src/knexQueryMapper.ts:104]: Is this changed related?

Comment (craicoverflow) [reply packages/graphback-runtime-knex/src/knexQueryMapper.ts:104]: I found a bug while testing the querying.

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts (hunk 2 above)
Comment (craicoverflow) [reply packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:294]: An unhandled error is thrown. Thanks.

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts (hunk 1 above)
Comment (craicoverflow) [reply packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:298]:
```
"""@model"""
type Note {
  id: ID!
  comments: [Comment]!
}
```

`Note.comments` will need to map to a newly created input type, so this ensures that the required level is the exact same in `CreateNoteInput.comments` as `Note.comments`.

```
type CreateNoteInput {
  id: ID!
  comments: [CreateCommentInput]!
}
```

File: packages/graphback-runtime-knex/src/knexQueryMapper.ts (hunk 3 above)
Comment (machi1990) [reply packages/graphback-runtime-knex/src/knexQueryMapper.ts:104]: Nice catch. What was it?

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts (hunk 1 above)
Comment (machi1990) [reply packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:302]: Gotcha.

File: templates/ts-apollo-postgres-backend/model/datamodel.graphql [hunk 4]
```diff
@@ -4,14 +4,19 @@ type Note {
...
-  text: String
-  description: String
+  text: String!
+  description: String!
+  test: Test!
+  test2: [Test!]!
+}
+
+type Test {
+  hello: String
```
Comment (machi1990) [inline templates/ts-apollo-postgres-backend/model/datamodel.graphql:21]: Maybe we could add an [integration test](https://github.com/aerogear/graphback/tree/9622f7373cb59f3926b6a538ea4424de29af7d4e/integration) and revert the change on this template file?

File: packages/graphback-runtime-knex/src/knexQueryMapper.ts (hunk 3 above)
Comment (craicoverflow) [reply packages/graphback-runtime-knex/src/knexQueryMapper.ts:104]: `ne` was being evaluated as `eq`, so {ne: null} in SQL gave `IS NULL`.

Added https://github.com/aerogear/graphback/pull/1574/commits/0ac97c2d25774a468f08e243c67d81dbfdab712b to cover it.

File: templates/ts-apollo-postgres-backend/model/datamodel.graphql (hunk 4 above)
Comment (craicoverflow) [reply templates/ts-apollo-postgres-backend/model/datamodel.graphql]: Reverted file now, thanks. Have begun the process of adding integration test cases but not sure if I will get it complete in time (off tomorrow).

**Comment by machi1990** [comment]
Tested this issue and it worked out of the box for Mongo. But I ran into issues for Postgres. 

- Will a custom type as an array  e.g `comments: [Comment]` work ? I remember having issues with storing arrays using the `json` Postgres type and had to store arrays as `{data: array}`
- Another issue which is a general one is:

with the model below 
```
"""
@model
"""
type Note {
  id: ID!
  title: String!
  description: String
  """
  @db(type: 'json')
  """
  comments: [Comment]!
}

type Comment {
  text: String
  description: String
}

```
 
If I issue the following create mutation
```graphql
mutation {
  createNote(
    input: { title: "test", description: "description", comments: [{text: "mambo"}] }
  ) {
    comments {
      text
      description
    }
    title
  }
}

```

I had the following error 
```log

{
  "errors": [
    {
      "message": "insert into \"note\" (\"comments\", \"description\", \"title\") values ($1, $2, $3) returning \"comments\", \"title\" - invalid input syntax for type json",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "createNote"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "length": 171,
          "name": "error",
          "severity": "ERROR",
          "code": "22P02",
          "detail": "Expected \":\", but found \"}\".",
          "where": "JSON data, line 1: {\"{\\\"text\\\":\\\"mambo\\\"}\"}",
          "file": "json.c",
          "line": "1199",
          "routine": "report_parse_error",
          "stacktrace": [
            "error: insert into \"note\" (\"comments\", \"description\", \"title\") values ($1, $2, $3) returning \"comments\", \"title\" - invalid input syntax for type json",
            "    at Parser.parseErrorMessage 

```

**Comment by craicoverflow** [comment]
Thanks for spotting that @machi1990 - it should be fixed now with 2b601d870d7751a5ac526482efbfbfece7b0c114

File: packages/graphback-core/src/db/dataMapper.ts [hunk 5]
```diff
@@ -29,10 +30,19 @@ function getTableId(idField: string, data: any): TableID {
...
-  //TODO: Map fields to custom db names
-
   return {
     idField: getTableId(idField, data),
-    data
+    // eslint-disable-next-line @typescript-eslint/tslint/config
+    data: Object.keys(data).reduce((dataObj, key) => {
+      const value = data[key]
+
+      if (typeof value === 'object' || Array.isArray(value)) {
```
Comment (machi1990) [inline packages/graphback-core/src/db/dataMapper.ts:39]:
```suggestion
      if (typeof value === 'object') {
```

typeof value will aways be 'object' if it is an object or an array.

**Comment by machi1990** [comment]
> Thanks for spotting that @machi1990 - it should be fixed now with [2b601d8](https://github.com/aerogear/graphback/commit/2b601d870d7751a5ac526482efbfbfece7b0c114). Knex did not handle insertion of an array into a JSON field (it worked fine for objects), so I just had to stringify the data in advance of insert.

Cool thanks. I'll verify tomorrow but that should fix the issue too. I'll a full verification and full review tomorrow.

File: integration/tests/runtime-workflow-mongo.ts [hunk 6]
```diff
@@ -42,8 +42,8 @@ beforeAll(async () => {
     graphbackApi = buildGraphbackAPI(modelText, {
       dataProviderCreator: createMongoDbProvider(db),
       plugins: [
-        new SchemaCRUDPlugin({outputPath: "./output-mongo/schema/schema.graphql"}),
-        new ClientCRUDPlugin({format: 'graphql', outputFile: './output-mongo/client/graphback.graphql'})
+        new SchemaCRUDPlugin({ outputPath: "./output-mongo/schema/schema.graphql" }),
```
Comment (wtrocki) [inline integration/tests/runtime-workflow-mongo.ts:45]: This should be gone. We do not need schema plugin

Comment (machi1990) [reply integration/tests/runtime-workflow-mongo.ts:45]: It is needed, otherwise the integration tests fails with 
```
  ● Note 1 should be defined

    expect(received).toEqual(expected) // deep equality

    Expected: {"comments": [{"description": "Note A Comment Description", "id": "5ef5b6951f82b90668f8da94", "text": "Note A Comment"}, {"description": "Note A Comment Description", "id": "5ef5b6951f82b90668f8da95", "text": "Note A Comment 2"}], "description": "Note A Description", "id": "5ef5b6951f82b90668f8da90", "title": "Note A"}
    Received: null

      287 |   expect(response.data).toBeDefined();
      288 |   const notes = response.data.getNote;
    > 289 |   expect(notes).toEqual({
          |                 ^
      290 |     id: notesId[0],
      291 |     title: 'Note A',
      292 |     description: 'Note A Description',

      at tests/runtime-workflow-mongo.ts:289:17
      at fulfilled (../node_modules/tslib/tslib.js:112:62)

  ● Should create a new Note

    expect(received).toBeDefined()

    Received: undefined

      375 | test('Should create a new Note', async () => {
      376 |   const response = await createNote(client, { title: 'New note', description: 'New note description' });
    > 377 |   expect(response.data).toBeDefined();
          |                         ^
      378 |   expect(response.data.createNote).toEqual({ id: response.data.createNote.id, title: 'New note', description: 'New note description' });
      379 | })
      380 | 

      at tests/runtime-workflow-mongo.ts:377:25

```

File: packages/graphback-runtime-knex/src/knexQueryMapper.ts (hunk 3 above)
Comment (machi1990) [reply packages/graphback-runtime-knex/src/knexQueryMapper.ts:104]: Amazing thanks.

File: integration/tests/runtime-workflow-mongo.ts (hunk 6 above)
Comment (wtrocki) [reply integration/tests/runtime-workflow-mongo.ts:45]: Looks like we found bug that is not related to this PR

File: packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts (hunk 2 above)
Comment (machi1990) [reply packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:294]: Great. This works fine but it will fail when the user adds the `SchemaPlugin` to print out the schema 

```
    Error: Non-model NoteResultList.items maps to Graphback model (Note) which is not allowed.
        at /home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:296:15
        at Array.map (<anonymous>)
        at mapObjectInputFields (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:289:17)
        at Object.addCreateObjectInputType (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/definitions/schemaDefinitions.ts:319:13)
        at /home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts:731:11
        at Map.forEach (<anonymous>)
        at SchemaComposer.forEach (/home/manyanda/Documents/Projects/aerogear/graphback/node_modules/graphql-compose/lib/TypeStorage.js:37:23)
        at SchemaCRUDPlugin.createSchemaCRUDTypes (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts:724:20)
        at SchemaCRUDPlugin.buildSchemaForModels (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts:135:10)
        at SchemaCRUDPlugin.transformSchema (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts:63:10)
        at GraphbackPluginEngine.createSchema (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-core/src/plugin/GraphbackPluginEngine.ts:66:32)
        at GraphbackPluginEngine.createResources (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback-core/src/plugin/GraphbackPluginEngine.ts:50:10)
        at Object.buildGraphbackAPI (/home/manyanda/Documents/Projects/aerogear/graphback/packages/graphback/src/buildGraphbackAPI.ts:92:33)
        at /home/manyanda/Documents/Projects/aerogear/graphback/integration/tests/runtime-workflow-mongo.ts:42:20
        at Generator.next (<anonymous>)
        at fulfilled (/home/manyanda/Documents/Projects/aerogear/graphback/node_modules/tslib/tslib.js:112:62)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)

      at tests/runtime-workflow-mongo.ts:59:13

```

I ran into this issue locally when executing our integration tests - where we add the schema plugin to print out the schema so that it gets picked out by gql client
https://github.com/aerogear/graphback/blob/master/integration/tests/runtime-workflow-postgres.ts#L48-L50

But we already have a schema plugin defined by default https://github.com/aerogear/graphback/blob/master/packages/graphback/src/buildGraphbackAPI.ts#L82

/cc @wtrocki we kind of chatted about it today.

File: packages/graphback-core/src/db/dataMapper.ts [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import { ModelTableMap } from './buildModelTableMap';
+import { type } from 'os';
```
Comment (machi1990) [inline packages/graphback-core/src/db/dataMapper.ts:2]:
```suggestion
```

File: packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts [hunk 8]
```diff
@@ -682,14 +717,21 @@ export class SchemaCRUDPlugin extends GraphbackPlugin {
...
+    schemaComposer.add(JSONScalarInputType);
     schemaComposer.add(createInputTypeForScalar(GraphQLInt));
     schemaComposer.add(createInputTypeForScalar(GraphQLFloat));
 
     schemaComposer.forEach((tc: NamedTypeComposer<any>) => {
       const namedType = tc.getType();
-      if (isScalarType(namedType) && !isSpecifiedScalarType(namedType)) {
+      if (isScalarType(namedType) && !isSpecifiedScalarType(namedType) && !isSpecifiedGraphbackScalarType(namedType)) {
         schemaComposer.add(createInputTypeForScalar(namedType));
       }
```
Comment (machi1990) [inline packages/graphback-codegen-schema/src/SchemaCRUDPlugin.ts:728]:
```suggestion
      return;
      }
```

let's short-circuit the flow to avoid the checks below.

File: packages/graphback-core/src/plugin/getSelectedFieldsFromResolverInfo.ts [hunk 9]
```diff
@@ -21,6 +21,8 @@ export const getSelectedFieldsFromResolverInfo = (info: GraphQLResolveInfo, mode
 
         if (foundRelationship) {
           acc.push(foundRelationship.relationForeignKey)
+        } else {
```
Comment (machi1990) [inline packages/graphback-core/src/plugin/getSelectedFieldsFromResolverInfo.ts:24]: We'll need to check if it is not a `oneToMany` relationship before adding it.

Without it a relationship like this one below 
```
""" @model """
type Note {
  id: ID!
  title: String!
  description: String
  """
  @oneToMany(field: 'note')
  """
  comments: [Comment]!
}

""" @model """
type Comment {
  id: ID!
  text: String
  description: String
}
```

Would fail when issue a graphql request of this sort
```
findNotes {

 items {
  id, 
  comments {
    id
  }
}
}

```

it will fail consider `comments` as selectable field while it should not.

Comment (machi1990) [reply packages/graphback-core/src/plugin/getSelectedFieldsFromResolverInfo.ts:31]: Pushed a https://github.com/aerogear/graphback/commit/34365254d29a4146231169e1d409f998ab11968d to fix this issue and make postgres integration tests to pass. Working on the Mongo test cases now.

File: packages/graphback-core/src/db/dataMapper.ts [hunk 10]
```diff
@@ -29,10 +29,19 @@ function getTableId(idField: string, data: any): TableID {
 export const getDatabaseArguments = (modelMap: ModelTableMap, data?: any, fieldMap?: any): TableDataMap => {
   const idField = modelMap.idField;
 
-  //TODO: Map fields to custom db names
-
   return {
     idField: getTableId(idField, data),
-    data
+    // eslint-disable-next-line @typescript-eslint/tslint/config
+    data: Object.keys(data).reduce((dataObj, key) => {
```
Comment (machi1990) [inline packages/graphback-core/src/db/dataMapper.ts:35]: We have to check for if data is defined here. Otherwise it will crash. 

Since from https://github.com/aerogear/graphback/pull/1574/files#diff-5a1ded10109278ede19b3429e917d253R29 data can be undefined.

Comment (machi1990) [reply packages/graphback-core/src/db/dataMapper.ts:35]: I'll push a commit to only transform data if defined.

Comment (machi1990) [reply packages/graphback-core/src/db/dataMapper.ts:35]: This was the case for the MongoDB provider https://github.com/aerogear/graphback/blob/master/packages/graphback-runtime-mongodb/src/MongoDBDataProvider.ts#L130 causing the Mongo integration tests to fail.

**Comment by wtrocki** [comment]
No worries. I will do another verification soon

**Comment by wtrocki** [comment]
That will be actually very cool to have. Doing some workarounds for this feature in OVP at the moment

**Comment by wtrocki** [comment]
That will be actually very cool to have. Doing some workarounds for this feature in OVP at the moment

**Comment by wtrocki** [comment]
Would this work for scalars as well (along with the types)

**Comment by wtrocki** [comment]
Latest release doesn't seem to be working for querying data:
Screenshot 2020-06-28 at 22 06 34
Screenshot 2020-06-28 at 22 06 02

**Comment by machi1990** [comment]
> Latest release doesn't seem to be working for querying data:
> Screenshot 2020-06-28 at 22 06 34
> Screenshot 2020-06-28 at 22 06 02

I saw a fix for this in this PR.

**Comment by wtrocki** [comment]
Do we have integration test for this case?

**Comment by machi1990** [comment]
> Do we have integration test for this case?

I think the added integration tests should reflect that. 

The `tasks` custom type.

**Comment by machi1990** [comment]
@craicoverflow Thanks for merging this and for the amazing work.
